### ISSUE TITLE
GitHub Actions Migration - tracking issue

### ISSUE THREAD

**Issue by Waterdrips** [openfaas/faas#1585]
# Inlets
- [x] https://github.com/inlets/inlets-operator @jsiebens
- [x] https://github.com/inlets/inletsctl
- [ ] https://github.com/inlets/inlets

alexellis
- [x] k3sup @alexellis 
- [x] arkade @Waterdrips 
- [ ] https://github.com/alexellis/derek

# openfaas

Merged:
- [x] https://github.com/openfaas/faas-netes @akosveres
- [x] https://github.com/openfaas/templates @Waterdrips 
- [x] https://github.com/openfaas/nats-connector @CodeCutterUK 
- [x] https://github.com/openfaas/faasd @akosveres
- [x] https://github.com/openfaas/ofc-bootstrap @Waterdrips 
- [x] https://github.com/openfaas/faas-provider closed in https://github.com/openfaas/faas-provider/pull/49
- [x] https://github.com/openfaas/faas-cli
- [x] https://github.com/openfaas/kafka-connector
- [x] https://github.com/openfaas/cron-connector
- [x] https://github.com/openfaas/openfaas-cloud @Waterdrips 

Waiting on Review:
- [ ] https://github.com/openfaas/nats-queue-worker pending: https://github.com/openfaas/nats-queue-worker/pull/103
- [ ] https://github.com/openfaas/ingress-operator pending: https://github.com/openfaas/ingress-operator/pull/42
- [ ] https://github.com/openfaas/faas-idler pending https://github.com/openfaas/faas-idler/pull/44
- [ ] https://github.com/openfaas/faas-swarm pending: https://github.com/openfaas/faas-swarm/pull/76
- [ ] https://github.com/openfaas/certifier pending https://github.com/openfaas/certifier/pull/58
- [ ] https://github.com/openfaas/of-watchdog pending https://github.com/openfaas/of-watchdog/pull/114
- [ ] https://github.com/openfaas/connector-sdk pending https://github.com/openfaas/connector-sdk/pull/53
- [ ] https://github.com/openfaas/faas pending https://github.com/openfaas/faas/pull/1598

Pending:

- [ ] https://github.com/openfaas/store
- [ ] https://github.com/openfaas/mqtt-connector

# OpenFaaS Incubator 
- [ ] https://github.com/openfaas-incubator/openfaas-sqs-connector 
- [ ] https://github.com/openfaas-incubator/openfaas-vcenter-connector

**Comment by alexellis** [comment]
Folks unfortunately the first set of merged PRs did not work as I don't believe that we tested them in our branches with a release tag before submitting the PR and asking for a merge. Let's tighten up our work before asking for a merge?

## Make sure release tags work

I also noticed that the Docker v2 build/push action does not copy in a .git folder, this can be worked-around as follows (with a path and context). When these are missing, the code does a fresh clone / checkout of the code and lacks the .git folder used to get versions and tags.

```yaml
      - name: Build multi-arch containers for validation only
        uses: docker/build-push-action@v2
        with:
          context: .
          file: ./Dockerfile
          outputs: "type=image,push=false"
          platforms: linux/amd64,linux/arm/v7,linux/arm64
          tags: |
            ghcr.io/openfaas/faas-netes:${{ github.sha }}
```

## Include a NS variable in the image name

In order to test in our own repos, we should each have some way of overwriting `ghcr.io/openfaas/faas-netes:${{ github.sha }}` with our own username i.e. `ghcr.io/NS/faas-netes:${{ github.sha }}` where NS is a GitHub Action env-var. Have a look at https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables

## Keep naming consistent

build.yaml - upon * for push and branches, many PRs just assumed only master would be used
publish.yaml - upon tags

## Make sure this is multi-arch with buildx

We need to build for multiple architectures, not just `x86_64` - see the faas-netes build or the [other references examples on trello](https://trello.com/c/uqaun5lI/172-4th-nov-development-call).

**Comment by Waterdrips** [comment]
fyi you can get the repo name like this in an action
```
      - name: Get Repo
        id: get_repo
        run: echo ::set-output name=REPO::$(echo ${{ github.event.repository.owner.login }} | tr '[:upper:]' '[:lower:]')
```

then ref it with this
```
            ghcr.io/${{ steps.get_repo.outputs.REPO }}/${{ matrix.svc }}:latest
```

to allow pusing to a user's ghcr repo while testing

**Comment by CodeCutterUK** [comment]
@Waterdrips Yes but the catch is `event` isn't always included from what I've read, although that might only be for scheduled triggers....it's probably fine in the context of openfaas builds though, but not really my call.

**Comment by Waterdrips** [comment]
it looks like its only scheduled builds that dont have events from the docs and forums, we dont use those currently.

**Comment by CodeCutterUK** [comment]
So using this instead of `github.repository_owner` or the currently undocumented env `GITHUB_REPOSITORY_OWNER` results in the same issue if the repo owner string is storied in a secret, which is highly likely given that the GHCR username would like be stored as a secret. Again likely not an issue for repos in a 'proper' org like, but would likely be an issue for forks thus not fully solving the issue we wanted to solve.

You can see the issue here: https://github.com/CodeCutterUK/nats-connector/runs/1395097635?check_suite_focus=true
versus when the secrets are removed: https://github.com/CodeCutterUK/nats-connector/runs/1395105402?check_suite_focus=true

**Comment by Waterdrips** [comment]
I never found ${{ github.repository_owner }} - i have found the GH Actions docs to be not amazing... I like that more than using the event...
 

/so as far as i can tell the problem is: 

* images need all lowrcase repos.
* GHCR login needs your actual github username (With correct case) 

so we need both the explicit docker username, or use the non-lowercased repo owner (however we decide to get it) - need to decide how we manage this, we could use the ${{ github.repository_owner }} as the dockerhub login, and have a step that gets the lowercase version for image tags, repo path etc?

**Comment by Waterdrips** [comment]
@CodeCutterUK updated based on the above:https://github.com/Waterdrips/openfaas-cloud/actions/runs/361487319 seems to work well without the secret existing

**Comment by CodeCutterUK** [comment]
Yes, works perfectly without the lowercase version of the username as a secret :)

**Comment by developer-guy** [comment]
I completed the Github Action migration issue for the connector-sdk,  here is the link below:
https://github.com/openfaas/connector-sdk/pull/53

**Comment by alexellis** [comment]
Thanks to @cpanato who has volunteered to pick up the NATS queue worker. This will need a new PR, and I'd suggest using one of the other projects in this list as a template.
